frontends/PyCDE/src/pycde/module.py

```diff
@@ -45,45 +47,217 @@ class Input(ModuleDecl):
+      sys._generate_queue.append(self)
+    else:
+      paramdecl_list = [
+          hw.ParamDeclAttr.get_nodefault(i.name,
+                                         mlir.ir.TypeAttr.get(i.attr.type))
+          for i in self.parameters
+      ]
+      self.circt_mod = hw.HWModuleExternOp(
+          symbol,
+          self.input_ports,
+          self.output_ports,
+          parameters=paramdecl_list,
+          attributes={"verilogName": mlir.ir.StringAttr.get(self.extern_name)},
+          loc=self.loc,
+          ip=sys._get_ip())
+    self.add_accessors()
 
-  def __init__(self, value=None, name: str = None):
-    if value is not None:
-      self.attr = var_to_attribute(value)
-    self.name = name
+  @property
+  def is_created(self):
+    return self.circt_mod is not None
+
+  def instantiate(self, instance_name: str, inputs: dict, loc):
+    """Create a instance op."""
+    if self.extern_name is None:
+      return self.circt_mod.create(instance_name, **inputs, loc=loc)
+    else:
+      return self.circt_mod.create(instance_name,
+                                   **inputs,
+                                   parameters=self.parameters,
+                                   loc=loc)
+
+  def generate(self):
+    """Fill in (generate) this module. Only supports a single generator
+    currently."""
+    assert len(self.generators) == 1
+    for g in self.generators.values():
+      g.generate(self)
+      return
 
 
 # Set an input to no_connect to indicate not to connect it. Only valid for
 # external module inputs.
 no_connect = object()
 
 
-class module:
-  """Decorator for module classes or functions which parameterize module
-  classes."""
+def module(func_or_class):
+  """Decorator to signal that a class should be treated as a module or a
+  function should be treated as a module parameterization function. In the
+  latter case, the function must return a python class to be treated as the
+  parameterized module."""
+  if inspect.isclass(func_or_class):
+    # If it's just a module class, we should wrap it immediately
+    return _module_base(func_or_class, None)
+  elif inspect.isfunction(func_or_class):
+    return _parameterized_module(func_or_class, None)
+  raise TypeError(
+      "@module decorator must be on class or parameterization function")
+
+
+def _get_module_cache_key(func,
+                          params) -> Tuple[builtins.function, mlir.ir.DictAttr]:
+  """The "module" cache is specifically for parameterized modules. It maps the
+  module parameterization function AND parameter values to the class which was
+  generated by a previous call to said module parameterization function."""
+  if not isinstance(params, mlir.ir.DictAttr):
+    params = var_to_attribute(params)
+  return (func, mlir.ir.Attribute(params))
+
+
+# A memoization table for module parameterization function calls.
+_module_cache: typing.Dict[Tuple[builtins.function, mlir.ir.DictAttr],
```

Comment:
maybe put this global variable at the top of the file and name it like MODULE_CACHE?